### PR TITLE
add kafka limit push down

### PR DESCRIPTION
**Pull request by motto1314**
#7071

### CODE REVIEW

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java [hunk 1]
```diff
@@ -247,6 +282,34 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
             return Optional.empty();
         }
 
+        // FilterNode is not added, only if all domain match the following conditions:
```
Comment (Praveen2112) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java:285]: Can we handle them in a separate PR ?

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java [hunk 2]
```diff
@@ -140,10 +142,17 @@ public KafkaFilteringResult getKafkaFilterResult(
 
             // push down partitions
             final Set<Long> finalPartitionIdsFiltered = partitionIdsFiltered;
-            List<PartitionInfo> partitionFilteredInfos = partitionInfos.stream()
+            partitionInfos = partitionInfos.stream()
```
Comment (Praveen2112) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:145]: Why this change ?

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java [hunk 3]
```diff
@@ -237,6 +246,32 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
         return new ConnectorTableProperties();
     }
 
+    @Override
+    public Optional<LimitApplicationResult<ConnectorTableHandle>> applyLimit(ConnectorSession session, ConnectorTableHandle table, long limit)
```
Comment (Praveen2112) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java:250]: I guess we miss a few tests related to limit pushdown

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java [hunk 4]
```diff
@@ -140,10 +142,17 @@ public KafkaFilteringResult getKafkaFilterResult(
...
                     .collect(toImmutableList());
-            return new KafkaFilteringResult(partitionFilteredInfos, partitionBeginOffsets, partitionEndOffsets);
+        }
+
+        // push down limit
+        if (limit.isPresent()) {
+            Map<TopicPartition, Long> finalPartitionBeginOffsets = partitionBeginOffsets;
+            partitionEndOffsets = overridePartitionEndOffsets(partitionEndOffsets,
+                    partition -> (finalPartitionBeginOffsets.get(partition) != INVALID_KAFKA_RANGE_INDEX)
+                            ? Optional.of(finalPartitionBeginOffsets.get(partition) + limit.getAsLong()) : Optional.empty());
```
Comment (Praveen2112) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:155]: Please correct me if I am wrong.. this would work only if the each message is on size 1 byte then we might assume that if we add `limit` to it... If each message is of different sizes then the offset might differ

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java [hunk 5]
```diff
@@ -40,6 +40,7 @@
 import java.util.List;
```
Comment (kokosing) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:40]: @motto1314 Welcome in the community! Thank you for your contribution.

Can you please make sure you read and follow:
 - https://github.com/trinodb/trino/blob/master/CONTRIBUTING.md - we are not allowed to accept any change without CLA signed.
 - https://github.com/trinodb/trino/blob/master/DEVELOPMENT.md
 
 Can you please improve your commit message and explain add a better commit message and pull request title and description. Please mention what is the expected user experience change. 

It looks like you are adding a new functionality, can you please also add proper testing of that feature to `TestKafkaIntegrationSmokeTest`?

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java (hunk 2 above)
Comment (motto1314) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:145]: This change is to reuse the following，like line:157

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java (hunk 4 above)
Comment (motto1314) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:155]: Kafka offset is the metadata of Kafka and the inherent information of each message. 
According to my understanding, it should have nothing to do with the message size ? I don't understand the limitations of byte ?

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java (hunk 5 above)
Comment (motto1314) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:40]: thank you very much.
1、CLA has been sent to cla@trino.io , please check
2、DEVELOPMENT.md I will study it carefully
3、I'll consider adding a proper testing

**Comment by motto1314** [comment]
`select _partition_id,_partition_offset from TOPIC_TABLE where _partition_id=0 and _partition_offset>10 limit 10;`
This operation is expected to pass _partition_id and _partition_offset reads data by timing increment, and uses LIMIT as Max batchsize.
Expect to get:
```
  _partition_id | _partition_offset 
---------------+-------------------
             0 |                11 
             0 |                12 
             0 |                13 
             0 |                14 
             0 |                15 
             0 |                16 
             0 |                17 
             0 |                18 
             0 |                19 
             0 |                20 
(10 rows)
```
But actually:

```
_partition_id | _partition_offset 
---------------+-------------------
             0 |           4200011 
             0 |           4200012 
             0 |           4200013 
             0 |           4200014 
             0 |           4200015 
             0 |           4200016 
             0 |           4200017 
             0 |           4200018 
             0 |           4200019 
             0 |           4200020 
(10 rows)
```
Using order by _partition_offset can be realized, but it involves full table scanning and is inefficient.

this is ETL，because need partition_id, partition_offset and maxBatchSize，So we need to be able to read the fixed batch orderly.
To solve this problem, we use limit push down.

In the modification, the push down of limit does not change much, because the existence of LIMITNODE is retained.
The most controversial part is applyFilter, because in some scenarios, the original FilterNode will be deleted.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java (hunk 1 above)
Comment (motto1314) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java:285]: Do you mean to submit the applyfilter part separately？

**Comment by findepi** [comment]
@motto1314 you probably have already done this, but making sure -- did you file the CLA?

**Comment by motto1314** [comment]
> @motto1314 you probably have already done this, but making sure -- did you file the CLA?

Yes, I have followed the following requirements：
complete and sign, then scan and email a pdf file of this Agreement
to cla@prestosql.io

Did you not receive the email, or did you not see the attachment in the email？

**Comment by martint** [comment]
@cla-bot check

File: plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaIntegrationSmokeTest.java [hunk 6]
```diff
@@ -627,4 +627,35 @@ public String toString()
             return tableName; // for test case label in IDE
         }
     }
+
+    @Test
+    public void testApplyFilterAndLimit()
+    {
+        long partitionNum = (long) computeScalar("SELECT count(distinct _partition_id) partition_num FROM kafka.tpch.customer");
```
Comment (kokosing) [inline plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaIntegrationSmokeTest.java:634]: Why `partitionNum` can have different values? How many times this test is executed?

File: plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaIntegrationSmokeTest.java [hunk 7]
```diff
@@ -627,4 +627,35 @@ public String toString()
...
+
+        assertQuery("SELECT _partition_id, _partition_offset" +
```
Comment (kokosing) [inline plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaIntegrationSmokeTest.java:636]: This test is fine to see that there is no regression, but on the other hand there is no prove that limit pushdown actually worked. Can you please use `io.trino.sql.query.QueryAssertions.QueryAssert#isFullyPushedDown` to verify that.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java [hunk 8]
```diff
@@ -247,6 +282,34 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
             return Optional.empty();
         }
 
+        // FilterNode is not added, only if all domain match the following conditions:
+        // 1. columnName is PARTITION_ID_FIELD
+        // 2. columnName is PARTITION_OFFSET_FIELD, and offset is a singleValue or continuous range (eg: _partition_offset > 0)
+        Map<ColumnHandle, Domain> incomplete = new HashMap<>();
+        if (constraint.getSummary().getDomains().isPresent()) {
```
Comment (kokosing) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java:289]: Why do you need to change this for limit pushdown?

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java [hunk 9]
```diff
@@ -140,10 +142,17 @@ public KafkaFilteringResult getKafkaFilterResult(
...
+            partitionInfos = partitionInfos.stream()
                     .filter(partitionInfo -> finalPartitionIdsFiltered.contains((long) partitionInfo.partition()))
                     .collect(toImmutableList());
-            return new KafkaFilteringResult(partitionFilteredInfos, partitionBeginOffsets, partitionEndOffsets);
+        }
+
+        // push down limit
+        if (limit.isPresent()) {
+            Map<TopicPartition, Long> finalPartitionBeginOffsets = partitionBeginOffsets;
+            partitionEndOffsets = overridePartitionEndOffsets(partitionEndOffsets,
```
Comment (kokosing) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:153]: nit: put each argument in separate line.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java [hunk 10]
```diff
@@ -140,10 +142,17 @@ public KafkaFilteringResult getKafkaFilterResult(
...
+                    partition -> (finalPartitionBeginOffsets.get(partition) != INVALID_KAFKA_RANGE_INDEX)
```
Comment (kokosing) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:154]: use regular `if` instead of `?` operator

**Comment by bitsondatadev** [comment]
:wave: @motto1314 - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.
